File: 13.SlidingWindow滑动窗口/leetcode_1658_minOperations.cpp
#include <iostream>
#include <vector>
#include <climits>
using namespace std;

class Solution {
public:
    int minOperations(vector<int>& nums, int x) {
        x = -x;
        for (int i : nums) x+=i;
        int r = 0;
        int tot = 0;
        int res = INT_MAX;
        int n = nums.size();
        for (int l=0; l<n; l++){
            while (r<n && tot<x){
                tot += nums[r];
                r++;
            }
            if (tot == x && l <= r){
                res = min(res, n-(r-l));
            }
            tot -= nums[l];
        }
        return res==INT_MAX ? -1 : res;
    }
};

int main(){
    Solution s;
    vector<int> nums = {1,1,4,2,3};
    int x=5;
    cout << s.minOperations(nums, x) << endl;
}

File: 11.PrefixSum前缀和/leetcode_1658_minOperations.py
from collections import defaultdict

class Solution:
    def minOperations(self, nums: list[int], x: int) -> int:
        n = len(nums)
        preCnt = defaultdict(int)
        sufCnt = defaultdict(int)
        preCnt[0] = sufCnt[0] = 0
        pre, suf = 0, 0
        for i in range(n):
            pre += nums[i]
            suf += nums[n-i-1]
            preCnt[pre] = i+1
            sufCnt[suf] = i+1
        
        res = 10**18
        for k,v in preCnt.items():
            if x-k in sufCnt and sufCnt[x-k]+v <= n:
                res = min(res, sufCnt[x-k]+v)
        return -1 if res==10**18 else res

if __name__ == "__main__":
    s = Solution()
    nums = [1,1,4,2,3]
    x = 5
    print(s.minOperations(nums, x))
